### PR TITLE
Services API

### PR DESCRIPTION
**Pull request by jerone**
Fixes: https://github.com/suda/toolbar/issues/24

* [x] :sparkles: Implement [services API](https://atom.io/docs/latest/behind-atom-interacting-with-packages-via-services)
* [x] :package: Increase atom engine semver required for services API
* [x] :art: Replace `(append|prepend)(Button|Spacer)` with `addButton()` and `addSpacer()` API
* [x] :art: Keep legacy API for now, use [grim](https://www.npmjs.com/package/grim) (see screenshot below)
* [x] :art: Group buttons & spacers for removing
* [x] :sparkles: Implement `removeItems()` API by using groups (partially fixes https://github.com/suda/toolbar/issues/1)
* [x] :art: Make button options an object
* [x] :sparkles: Add priority option to position a button (defaults to highest already added priority, with a start-priority of `50`)
* [x] :package: Require `space-pen` instead of `atom-space-pen-views`
* [x] :art: Made buttons/spacers into views, same as StatusBar does
* [x] :memo: Update README with new service API and new options

~~This PR is work-in-progress, not merge-ready.~~ Feature-complete & merge-ready.
This PR needs full testing. Please help with this.

:bulb: Code is written to support multiple toolbars one day.
:bulb: It is also possible for a future PR to make all priorities below zero to align right/bottom.

Use-case with new service API: https://github.com/cakecatz/flex-toolbar/pull/12
Use-case with legacy API: https://github.com/cakecatz/flex-toolbar & https://atom.io/packages/toolbar-main

Screenshot Deprecation Cop of legacy API usage:
deprecation

Legacy code should be removed with next major/minor package update.

### CODE REVIEW

**Comment by jerone** [comment]
@suda @cakecatz Do you guys have time to review and test this PR ?

**Comment by suda** [comment]
@jerone on it :)

File: README.md [hunk 1]
```diff
@@ -25,49 +25,81 @@ This package provides extensible toolbar for Atom.
...
-      alert 'foo'
-    , 'Show Alert'
-```
+In `package.json`:
 
-You can also prepend buttons/spacers using `prependButton()` and `prependSpacer()` methods.
+```json
+"package-dependencies": {
+  "toolbar": "^0.1.0"
+},
```
Comment (jerone) [inline README.md]: Above three lines are actually redundant without https://github.com/travs/atom-package-dependencies which probably shouldn't be used until it's merged into Atom core: https://github.com/atom/apm/issues/288

File: package.json [hunk 2]
```diff
@@ -6,13 +6,23 @@
...
+    "atom": ">=0.188.0"
   },
   "dependencies": {
-    "atom-space-pen-views": "^2.0.3",
+    "grim": "^1.2.2",
+    "space-pen": "^5.1.1",
     "underscore-plus": "^1.6.6"
+  },
+  "providedServices": {
+    "tool-bar": {
```
Comment (jerone) [inline package.json]: The service is called `tool-bar` same as https://github.com/atom/status-bar/blob/master/package.json#L17

Comment (suda) [inline package.json]: Yeah, I think this package should be renamed to `tool-bar` too.

**Comment by suda** [comment]
`ToolbarManager` removed `(append|prepend)(Button|Spacer)`. Could it still provide those methods but with `grim` warnings?

File: package.json (hunk 2 above)
Comment (jerone) [inline package.json]: Wondering what the consequences are of renaming the whole package...

**Comment by jerone** [comment]
[**@suda**](https://github.com/suda) commented on [2 mei 2015 15:55 CEST](https://github.com/suda/toolbar/pull/34#issuecomment-98361170 "2015-05-02T13:55:55Z - Replied by Github Comment Enhancer"):
> `ToolbarManager` removed `(append|prepend)(Button|Spacer)`. Could it still provide those methods but with `grim` warnings?

They are: https://github.com/jerone/toolbar/blob/services-api/lib/toolbar.coffee#L34-L59

File: package.json [hunk 3]
```diff
@@ -6,13 +6,23 @@
...
+      "description": "Does a useful thing",
```
Comment (jerone) [inline package.json]: TODO: insert useful message.

**Comment by suda** [comment]
Aaah right. I was using code from example:

```
consumeToolBar: (toolbar) ->
  @toolbar = toolbar 'another-toolbar'

  # Adding button
  @toolbar.appendButton
    icon: 'octoface'
    callback: 'application:about'
    tooltip: 'About Atom'
```
which tries to call `appendButton` on manager, not toolbar.

File: package.json (hunk 3 above)
Comment (suda) [inline package.json]: How about "A container for buttons at workspace edge"?

**Comment by jerone** [comment]
[**@suda**](https://github.com/suda) commented on [2 mei 2015 16:03 CEST](https://github.com/suda/toolbar/pull/34#issuecomment-98362172 "2015-05-02T14:03:19Z - Replied by Github Comment Enhancer"):
> Aaah right. I was using code from example:
> 
>     consumeToolBar: (toolbar) -&gt;
>       @toolbar = toolbar 'another-toolbar'
> 
>       # Adding button
>       @toolbar.appendButton
>         icon: 'octoface'
>         callback: 'application:about'
>         tooltip: 'About Atom'
> 
> which tries to call `appendButton` on manager, not toolbar.

Ah good catch, will update right now!

File: package.json (hunk 3 above)
Comment (jerone) [inline package.json]: To keep inline with StatusBar, yes. Will update now.

File: lib/toolbar-view.coffee [hunk 4]
```diff
@@ -1,14 +1,36 @@
 {CompositeDisposable} = require 'atom'
-{View} = require 'atom-space-pen-views'
+{View, $} = require 'space-pen'
```
Comment (jerone) [inline lib/toolbar-view.coffee]: Remove redundant `$`.

**Comment by suda** [comment]
@jerone Ok, I think it can be merged. Do you want to add anything more before?

**Comment by jerone** [comment]
[**@suda**](https://github.com/suda) commented on [2 mei 2015 16:59 CEST](https://github.com/suda/toolbar/pull/34#issuecomment-98368312 "2015-05-02T14:59:14Z - Replied by Github Comment Enhancer"):
> [@jerone](https://github.com/jerone) Ok, I think it can be merged. Do you want to add anything more before?

Oke, last commit done.

You only have to come up with a correct semver and match that [here](https://github.com/jerone/toolbar/blob/services-api/package.json#L4), [here](https://github.com/jerone/toolbar/blob/services-api/package.json#L24) and [here](https://github.com/jerone/toolbar/blob/services-api/README.md) (two times).

**Comment by suda** [comment]
I would say it's a great moment to bump it to `0.1.0` :) 🎉🎊

File: package.json (hunk 2 above)
Comment (suda) [inline package.json]: Renamed to `tool-bar`. Both [GH](https://github.com/suda/toolbar) and [Atom](https://atom.io/packages/toolbar) redirect to new address.

Comment (jerone) [inline package.json]: Cool. Didn't know it was that easy.

Comment (suda) [inline package.json]: When publishing, you have to leave `name` in `package.json` as was before and then do:
`$ apm publish VERSION --rename NEW_NAME` :)

Comment (jerone) [inline package.json]: TIL

Comment (jerone) [inline package.json]: hmmm, I don't think apm is ready for renamed packages. I updated the packages via Atom available update (thus not via git) and I got two tool-bars and two times the main tool-bar (see screenshot). I also got folders two of both packages (see screenshot too). And lost my settings (ain't really bad, as it is only two settings).

tool-bar

I think it's an acceptable change, but now you know.

**Pull request merged**